crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+    fn decode(
+        value: <Postgres as sqlx::Database>::ValueRef<'r>,
+    ) -> Result<Self, sqlx::error::BoxDynError> {
+        <i64 as sqlx::Decode<Postgres>>::decode(value).map(|milliseconds_since_epoch| {
+            SecondsTimestamp(DateTime::from_timestamp_nanos(
+                milliseconds_since_epoch * 1_000_000_000,
+            ))
+        })
+    }
+}
+
+impl sqlx::Type<Postgres> for SecondsTimestamp {
+    fn type_info() -> <Postgres as sqlx::Database>::TypeInfo {
+        <i64 as sqlx::Type<Postgres>>::type_info()
+    }
+}
+
+#[derive(Clone, Debug, FromRow)]
+#[enum_def(table_name = "users")]
+pub struct SynapseUser {
+    /// Full User ID of the user
+    pub name: FullUserId,
+    /// Password hash string for the user. Optional (null if no password is set).
+    pub password_hash: Option<String>,
+    /// Whether the user is a Synapse Admin
+    pub admin: SynapseBool,
+    /// Whether the user is deactivated
+    pub deactivated: SynapseBool,
+    /// When the user was created
+    pub creation_ts: SecondsTimestamp,
+    // TODO ...
+    // TODO is_guest
+    // TODO do we care about upgrade_ts (users who upgraded from guest accounts to real accounts)
+}
+
+#[derive(Iden)]
+pub enum ExtraSynapseUserIden {
+    AppserviceId,
+    IsGuest,
+}
+
+/// List of Synapse tables that we should acquire an `EXCLUSIVE` lock on.
+///
+/// This is a safety measure against other processes changing the data underneath our feet.
+/// It's still not a good idea to run Synapse at the same time as the migration.
+// TODO not complete!
+const TABLES_TO_LOCK: &[&str] = &["users"];
+
+/// Number of migratable rows in various Synapse tables.
+/// Used to estimate progress.
+#[derive(Clone, Debug)]
+pub struct SynapseRowCounts {
+    pub users: i64,
+}
+
+pub struct SynapseReader<'c> {
+    conn: &'c mut PgConnection,
+}
+
+impl<'conn> SynapseReader<'conn> {
+    /// Create a new Synapse reader, which entails creating a transaction and locking Synapse tables.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error
+    /// - If we can't lock the Synapse tables (pointing to the fact that Synapse may still be running)
+    pub async fn new(
+        synapse_connection: &'conn mut PgConnection,
+        dry_run: bool,
+    ) -> Result<Self, Error> {
+        query("BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;")
+            .execute(&mut *synapse_connection)
+            .await
+            .into_database("begin transaction")?;
+
+        let lock_type = if dry_run {
+            // We expect dry runs to be done alongside Synapse running, so we don't want to
+            // interfere with Synapse's database access in that case.
+            "ACCESS SHARE"
+        } else {
+            "EXCLUSIVE"
+        };
+        for table in TABLES_TO_LOCK {
+            query(&format!("LOCK TABLE {table} IN {lock_type} MODE NOWAIT;"))
```

Comment:
What exactly happens if we lock the table when Synapse is running? The `NOWAIT` means this will fail, or will Synapse just start hanging because it can't query those tables?